File: algorithms/cpp/maximumLengthofaConcatenatedStringwithUniqueCharacters/maximumLengthofaConcatenatedStringwithUniqueCharacters.cpp
// Source: https://leetcode.com/problems/maximum-length-of-a-concatenated-string-with-unique-characters/
// Author: Miao Zhang
// Date:   2021-04-19

class Solution {
public:
    int maxLength(vector<string>& arr) {
        vector<int> bitarr;
        for (string& a: arr) {
            set<char> s(begin(a), end(a));
            if (s.size() != a.size()) continue;
            bitarr.push_back(0);
            for (char c: a) bitarr.back() |= 1 << (c - 'a');
        }

        int res = 0;
        function<void(int, int)> dfs = [&] (int start, int cur) {
            res = max(res, __builtin_popcount(cur));
            for (int i = start; i < bitarr.size(); i++) {
                if ((cur & bitarr[i]) == 0) {
                    dfs(i + 1, cur | bitarr[i]);
                }
            }
        };

        dfs(0, 0);
        return res;
    }
};

File: algorithms/python/maximumLengthofaConcatenatedStringwithUniqueCharacters/maximumLengthofaConcatenatedStringwithUniqueCharacters.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/maximum-length-of-a-concatenated-string-with-unique-characters/
# Author: Miao Zhang
# Date:   2021-04-19

class Solution:
    def maxLength(self, arr: List[str]) -> int:
        setarr = []
        for a in arr:
            if len(set(a)) == len(a):
                setarr.append(set(a))
        
        self.res = 0
        def dfs(start, cur):
            self.res = max(self.res, len(cur))
            for i in range(start, len(setarr)):
                if not (cur & setarr[i]):
                    dfs(i + 1, cur | setarr[i])
        
        dfs(0, set())
        return self.res
